src/Ens.ts

```diff
@@ -0,0 +1,594 @@
+        item[1] === currencyTicker.toUpperCase() ||
+        item[2] === currencyTicker.toUpperCase(),
+    );
+    if (coin < 0 || !formatsByCoinType[coin]) {
+      throw new ResolutionError(ResolutionErrorCode.UnsupportedCurrency, {
+        currencyTicker,
+      });
+    }
+
+    return coin.toString();
+  }
+
+  private fromUDRecordNameToENS(record: string): string {
+    const mapper = {
+      'ipfs.redirect_domain.value': 'url',
+      'browser.redirect_url': 'url',
+      'whois.email.value': 'email',
+      'gundb.username.value': 'gundb_username',
+      'gundb.public_key.value': 'gundb_public_key',
+    };
+    return mapper[record] || record;
+  }
+
+  private async addr(
+    domain: string,
+    currencyTicker: string,
+  ): Promise<string | undefined> {
+    const resolver = await this.resolver(domain).catch(
+      (err: ResolutionError) => {
+        if (err.code !== ResolutionErrorCode.UnspecifiedResolver) {
+          throw err;
+        }
+      },
+    );
+
+    const isWrappedDomain = await this.determineIsWrappedDomain(
+      this.namehash(domain),
+    );
+    if (isWrappedDomain) {
+      return await this.getAddressForWrappedDomain(domain);
+    }
+
+    if (!resolver) {
+      const owner = await this.owner(domain);
+      if (isNullAddress(owner)) {
+        throw new ResolutionError(ResolutionErrorCode.UnregisteredDomain, {
+          domain,
+        });
+      }
+      throw new ResolutionError(ResolutionErrorCode.UnspecifiedResolver, {
+        domain,
+      });
+    }
+
+    const coinType = this.getCoinType(currencyTicker.toUpperCase());
+    return await this.fetchAddress(resolver, domain, coinType);
+  }
+
+  private async fetchAddress(
+    resolver: string,
+    domain: string,
+    coinType: string,
+  ): Promise<string | undefined> {
+    const formatsByCoinType = requireOrFail(
+      '@ensdomains/address-encoder',
+      '@ensdomains/address-encoder',
+      '>= 0.1.x <= 0.2.x',
+    ).formatsByCoinType;
+
+    const resolverContract = new EthereumContract(
+      resolverInterface(resolver, coinType),
+      resolver,
+      this.provider,
+      this.proxyServiceApiKey,
+    );
+
+    const nodeHash = this.namehash(domain);
+    const addr: string =
+      coinType !== EthCoinIndex
+        ? await this.callMethod(resolverContract, 'addr', [nodeHash, coinType])
+        : await this.callMethod(resolverContract, 'addr', [nodeHash]);
+    if (isNullAddress(addr)) {
+      return undefined;
+    }
+
+    // eslint-disable-next-line no-undef
+    const data = Buffer.from(addr.replace('0x', ''), 'hex');
+    return formatsByCoinType[coinType].encoder(data);
+  }
+
+  private async getTextRecord(domain, key): Promise<string | undefined> {
+    const nodeHash = this.namehash(domain);
+    const resolver = await this.getResolverContract(domain);
+    return await this.callMethod(resolver, 'text', [nodeHash, key]);
+  }
+
+  private async getContentHash(domain: string): Promise<string | undefined> {
```

Comment:
from slack discussion, we should return all possible decode.
using this lib https://github.com/ensdomains/address-encoder

Reply:
Referring to this discussion: https://unstoppabledomains.slack.com/archives/C022W48Q10C/p1695220129816159

Reply:
What's the use of getting all decodes? Is this the same use for this function? 

I see this in the website code: https://github.com/unstoppabledomains/unstoppable-domains-website/blob/22045d5539478eda1e92f0aa255daf6f2e836778/backend/lib/blockchain/resolutionRecordsDecoder.ts#L47C1-L64C2

That thread and the discussion for me understanding what everyone trying to talk about:
![](https://i.imgur.com/T8C19go.gif)

Reply:
ContentHash is for a different thing then the address-encoder

ContentHash, see https://docs.ens.domains/ens-improvement-proposals/ensip-7-contenthash-field